theseus/labs/lie/functional/se3_impl.py

```diff
@@ -527,65 +361,111 @@ def _jlog_impl(group: torch.Tensor) -> Tuple[List[torch.Tensor], torch.Tensor]:
         (sine_theta + two_cosine_minus_two) / (theta2_nz * two_cosine_minus_two_nz),
     )
 
-    translation = group[:, :, 3].view(-1, 3, 1)
-    ret_lin = a.view(-1, 1) * group[:, :, 3]
-    ret_lin -= 0.5 * torch.cross(ret_ang, group[:, :, 3], dim=1)
-    ret_ang_ext = ret_ang.view(-1, 3, 1)
-    ret_lin += b.view(-1, 1) * (
-        ret_ang_ext @ (ret_ang_ext.transpose(1, 2) @ translation)
-    ).view(-1, 3)
+    translation = group[..., 3].view(*size, 3, 1)
+    ret_lin = a.view(*size, 1) * group[..., 3]
+    ret_lin -= 0.5 * torch.cross(ret_ang, group[..., 3], dim=-1)
+    ret_ang_ext = ret_ang.view(*size, 3, 1)
+    ret_lin += b.view(*size, 1) * (
+        ret_ang_ext @ (ret_ang_ext.transpose(-1, -2) @ translation)
+    ).view(*size, 3)
 
-    # jacobians
-    jac = group.new_zeros(group.shape[0], 6, 6)
+    return torch.cat([ret_lin, ret_ang], dim=-1), (
+        theta,
+        theta2,
+        theta2_nz,
+        sine,
+        cosine,
+        two_cosine_minus_two_nz,
+        a,
+        b,
+    )
 
-    b_ret_ang = b.view(-1, 1) * ret_ang
-    jac[:, :3, :3] = b_ret_ang.view(-1, 3, 1) * ret_ang.view(-1, 1, 3)
 
-    half_ret_ang = 0.5 * ret_ang
-    jac[:, 0, 1] -= half_ret_ang[:, 2]
-    jac[:, 1, 0] += half_ret_ang[:, 2]
-    jac[:, 0, 2] += half_ret_ang[:, 1]
-    jac[:, 2, 0] -= half_ret_ang[:, 1]
-    jac[:, 1, 2] -= half_ret_ang[:, 0]
-    jac[:, 2, 1] += half_ret_ang[:, 0]
+def _log_impl(group: torch.Tensor) -> torch.Tensor:
+    check_group_tensor(group)
+    ret, _ = _log_impl_helper(group)
+    return ret
 
-    diag_jac_rot = torch.diagonal(jac[:, :3, :3], dim1=1, dim2=2)
-    diag_jac_rot += a.view(-1, 1)
 
-    jac[:, 3:, 3:] = jac[:, :3, :3]
+def _jlog_impl_helper(
+    tangent_vector: torch.Tensor,
+    theta: torch.Tensor,
+    theta2: torch.Tensor,
+    theta2_nz: torch.Tensor,
+    sine: torch.Tensor,
+    cosine: torch.Tensor,
+    two_cosine_minus_two_nz: torch.Tensor,
+    a: torch.Tensor,
+    b: torch.Tensor,
+):
+    ret_lin = tangent_vector[..., :3]
+    ret_ang = tangent_vector[..., 3:]
+    size = tangent_vector.shape[:-1]
+    near_zero = theta < constants._SO3_NEAR_ZERO_EPS[tangent_vector.dtype]
+    non_zero = tangent_vector.new_ones(1)
+    jac = tangent_vector.new_zeros(*size, 6, 6)
+    jac[..., :3, :3], (b_ret_ang,) = SO3._jlog_impl_helper(ret_ang, theta, sine, cosine)
+    jac[..., 3:, 3:] = jac[..., :3, :3]
 
     theta_nz = torch.where(near_zero, non_zero, theta)
     theta4_nz = theta2_nz**2
     c = torch.where(
         near_zero,
         -1 / 360.0 - theta2 / 7560.0,
```

Comment:
Do we need `new_tensor()` here?

Reply:
I don't think so since `theta` is already a `torch.Tensor`.

Reply:
Oh sorry, I missed the tensor.